In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from pandas.api.types import CategoricalDtype


Caching the list of root modules, please wait!
(This will only be done once - type '%rehashx' to reset cache!)



# (Ignore) Combining the data files

In [ ]:
DATA_DIR = "data/"
# benign = pd.read_csv(DATA_DIR + "Benign Traffic.csv")
# ddos_icmp = pd.read_csv(DATA_DIR + "DDoS ICMP Flood.csv")
# ddos_udp = pd.read_csv(DATA_DIR + "DDoS UDP Flood.csv")
# dos_icmp = pd.read_csv(DATA_DIR + "DoS ICMP Flood.csv")
# dos_tcp = pd.read_csv(DATA_DIR + "DoS TCP Flood.csv")
# dos_udp = pd.read_csv(DATA_DIR + "DoS UDP Flood.csv")
# mitm_arp = pd.read_csv(DATA_DIR + "MITM ARP Spoofing.csv")
# ddos_publish = pd.read_csv(DATA_DIR + "MQTT DDoS Publish Flood.csv")
# dos_connect = pd.read_csv(DATA_DIR + "MQTT DoS Connect Flood.csv")
# dos_publish = pd.read_csv(DATA_DIR + "MQTT DoS Publish Flood.csv")
# mqtt_malformed = pd.read_csv(DATA_DIR + "MQTT Malformed.csv")
# os_scan = pd.read_csv(DATA_DIR + "Recon OS Scan.csv")
# ping_sweep = pd.read_csv(DATA_DIR + "Recon Ping Sweep.csv")
# port_scan = pd.read_csv(DATA_DIR + "Recon Port Scan.csv")
# vuln_scan = pd.read_csv(DATA_DIR + "Recon Vulnerability Scan.csv")

# # Combine all dataframes into a single dataframe and export to parquet
# all_data = pd.concat([
#     benign,
#     ddos_icmp,
#     ddos_udp,
#     dos_icmp,
#     dos_tcp,
#     dos_udp,
#     mitm_arp,
#     ddos_publish,
#     dos_connect,
#     dos_publish,
#     mqtt_malformed,
#     os_scan,
#     ping_sweep,
#     port_scan,
#     vuln_scan
# ], ignore_index=True)
# all_data.to_parquet(DATA_DIR + "combined_traffic.parquet", index=False)

# Initial Data Exploration

In [ ]:
data = pd.read_parquet(DATA_DIR + "combined_traffic.parquet")

In [ ]:
data["Timestamp"] = pd.to_datetime(data["Timestamp"], infer_datetime_format=True)
data["Attack Name"] = data["Attack Name"].astype("category")
attack_order = [
    "Benign",
    "DDoS ICMP Flood",
    "DDoS UDP Flood",
    "DoS ICMP Flood",
    "DoS TCP Flood",
    "DoS UDP Flood",
    "MITM ARP Spoofing",
    "MQTT DDoS Publish Flood",
    "MQTT DoS Connect Flood",
    "MQTT DoS Publish Flood",
    "MQTT Malformed",
    "Recon OS Scan",
    "Recon Ping Sweep",
    "Recon Port Scan",
    "Recon Vulnerability Scan"
]
cat_type = CategoricalDtype(categories=attack_order, ordered=True)
data["Attack Name"] = data["Attack Name"].astype(cat_type)
data["Attack Type"] = data["Attack Name"].cat.codes
# attack_mapping = {
#     "Benign": 0,
#     "DDoS ICMP Flood": 1,
#     "DDoS UDP Flood": 2,
#     "DoS ICMP Flood": 3,
#     "DoS TCP Flood": 4,
#     "DoS UDP Flood": 5,
#     "MITM ARP Spoofing": 6,
#     "MQTT DDoS Publish Flood": 7,
#     "MQTT DoS Connect Flood": 8,
#     "MQTT DoS Publish Flood": 9,
#     "MQTT Malformed": 10,
#     "Recon OS Scan": 11,
#     "Recon Ping Sweep": 12,
#     "Recon Port Scan": 13,
#     "Recon Vulnerability Scan": 14
# }
data.head()

/tmp/ipykernel_715498/2624595614.py:2: UserWarning: The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.
  data["Timestamp"] = pd.to_datetime(data["Timestamp"], infer_datetime_format=True)
/tmp/ipykernel_715498/2624595614.py:2: UserWarning: Parsing dates in %d/%m/%Y %I:%M:%S %p format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data["Timestamp"] = pd.to_datetime(data["Timestamp"], infer_datetime_format=True)


Flow ID         Src IP  Src Port  \
0     34.173.20.6-192.168.137.234-443-58644-6    34.173.20.6       443   
1     34.173.20.6-192.168.137.234-443-58646-6    34.173.20.6       443   
2  52.40.210.103-192.168.137.40-18665-60378-6  52.40.210.103     18665   
3            10.0.0.3-10.0.0.254-44505-1883-6       10.0.0.3     44505   
4            10.0.0.7-10.0.0.254-45727-1883-6       10.0.0.7     45727   

            Dst IP  Dst Port  Protocol           Timestamp  Flow Duration  \
0  192.168.137.234     58644         6 2023-09-14 09:24:26       14746364   
1  192.168.137.234     58646         6 2023-09-14 09:24:41       60102110   
2   192.168.137.40     60378         6 2023-09-14 09:24:14       90106769   
3       10.0.0.254      1883         6 2023-09-14 09:24:15      119998980   
4       10.0.0.254      1883         6 2023-09-14 09:24:15      119998539   

   Total Fwd Packet  Total Bwd packets  ...  Active Std  Active Max  \
0                 6                  0  ...         0.0         0.0   
1                18                  0  ...         0.0    219123.0   
2                 9                  0  ...         0.0    582529.0   
3               123                123  ...         0.0         0.0   
4               123                123  ...         0.0         0.0   

   Active Min   Idle Mean       Idle Std    Idle Max    Idle Min  Attack Name  \
0         0.0  14681116.0       0.000000  14681116.0  14681116.0          NaN   
1    219123.0  14957706.0  190543.248665  15061923.0  14672051.0          NaN   
2    582529.0  29841412.0  319090.538141  30049066.0  29473997.0          NaN   
3         0.0         0.0       0.000000         0.0         0.0          NaN   
4         0.0         0.0       0.000000         0.0         0.0          NaN   

   Label  Attack Type  
0      0           -1  
1      0           -1  
2      0           -1  
3      0           -1  
4      0           -1  

[5 rows x 86 columns]

In [ ]:
# Counts for each type of attack
data["Attack Name"].value_counts()

Attack Name
DoS TCP Flood               2106916
Recon Port Scan              485522
MQTT DDoS Publish Flood      413913
MQTT DoS Connect Flood       238031
Recon OS Scan                 85317
Benign Traffic                32620
Recon Vulnerability Scan       8321
DoS UDP Flood                  3115
DDoS UDP Flood                 2576
DDoS ICMP Flood                2552
MQTT Malformed                 2246
DoS ICMP Flood                 2107
MITM ARP Spoofing              1053
MQTT DoS Publish Flood          953
Recon Ping Sweep                 71
Name: count, dtype: int64

**Note**: Distribution for each type of attack & benign wildly varies. There's only 32620 benign, but orders of magnitude more for DoS TCP Flood. We may want to take subsets to have a better distribution and not have crazy class imbalance which can cause issues with biased predictions. 

In [ ]:
# No missing values
data.isna().sum().sum()

np.int64(0)

In [22]:
not_useful_columns = [
    "Flow ID",
    "Src IP",
    "Dst IP",
    "Attack Name",
]

In [19]:
# Datatypes of each column
pd.set_option('display.max_rows', None)  # To display all rows if there are many columns
data.dtypes

Flow ID                               object
Src IP                                object
Src Port                               int64
Dst IP                                object
Dst Port                               int64
Protocol                               int64
Timestamp                     datetime64[ns]
Flow Duration                          int64
Total Fwd Packet                       int64
Total Bwd packets                      int64
Total Length of Fwd Packet           float64
Total Length of Bwd Packet           float64
Fwd Packet Length Max                float64
Fwd Packet Length Min                float64
Fwd Packet Length Mean               float64
Fwd Packet Length Std                float64
Bwd Packet Length Max                float64
Bwd Packet Length Min                float64
Bwd Packet Length Mean               float64
Bwd Packet Length Std                float64
Flow Bytes/s                         float64
Flow Packets/s                       float64
Flow IAT M

In [ ]:
# corr_matrix = data.drop(not_useful_columns, axis=1).corr()
corr_matrix = pd.read_csv(DATA_DIR + "correlation_matrix.csv", index_col=0)

Src Port  Dst Port  Protocol  Timestamp  \
Src Port                    1.000000 -0.402987       NaN  -0.142296   
Dst Port                   -0.402987  1.000000       NaN   0.517558   
Protocol                         NaN       NaN       NaN        NaN   
Timestamp                  -0.142296  0.517558       NaN   1.000000   
Flow Duration               0.250639 -0.557067       NaN  -0.331726   
Total Fwd Packet            0.046628 -0.075297       NaN  -0.054052   
Total Bwd packets           0.006619 -0.003399       NaN   0.009746   
Total Length of Fwd Packet  0.010090 -0.008985       NaN  -0.000688   
Total Length of Bwd Packet -0.015181  0.013538       NaN  -0.000746   
Fwd Packet Length Max       0.090031 -0.082516       NaN  -0.027470   
Fwd Packet Length Min       0.059994 -0.056800       NaN  -0.034595   
Fwd Packet Length Mean      0.089277 -0.083647       NaN  -0.044878   
Fwd Packet Length Std       0.080545 -0.073333       NaN  -0.020742   
Bwd Packet Length Max      -0.138095  0.131399       NaN  -0.060229   
Bwd Packet Length Min      -0.076039  0.072660       NaN  -0.033187   
Bwd Packet Length Mean     -0.130094  0.124009       NaN  -0.059375   
Bwd Packet Length Std      -0.126814  0.120340       NaN  -0.056537   
Flow Bytes/s                0.000733 -0.000304       NaN  -0.003664   
Flow Packets/s              0.014555 -0.014290       NaN  -0.017020   
Flow IAT Mean               0.202354 -0.455705       NaN  -0.214554   
Flow IAT Std                0.136199 -0.274941       NaN  -0.082368   
Flow IAT Max                0.170093 -0.359926       NaN  -0.150760   
Flow IAT Min                0.184087 -0.394626       NaN  -0.211710   
Fwd IAT Total               0.249345 -0.556248       NaN  -0.339891   
Fwd IAT Mean                0.199664 -0.465123       NaN  -0.234683   
Fwd IAT Std                 0.128907 -0.261931       NaN  -0.070496   
Fwd IAT Max                 0.169201 -0.363192       NaN  -0.168827   
Fwd IAT Min                 0.186518 -0.423111       NaN  -0.285731   
Bwd IAT Total               0.058401 -0.082600       NaN   0.289377   
Bwd IAT Mean                0.048238 -0.062378       NaN   0.199890   
Bwd IAT Std                 0.044170 -0.062403       NaN   0.170117   
Bwd IAT Max                 0.047342 -0.064942       NaN   0.204112   
Bwd IAT Min                 0.021007 -0.020861       NaN   0.088345   
Fwd PSH Flags               0.063029 -0.079618       NaN   0.218135   
Bwd PSH Flags                    NaN       NaN       NaN        NaN   
Fwd URG Flags               0.003974 -0.002807       NaN   0.007459   
Bwd URG Flags                    NaN       NaN       NaN        NaN   
Fwd Header Length           0.034808 -0.052362       NaN  -0.032687   
Bwd Header Length           0.006922 -0.004322       NaN   0.008162   
Fwd Packets/s               0.016105 -0.015783       NaN  -0.016078   
Bwd Packets/s               0.001913 -0.002004       NaN  -0.003482   
Packet Length Min           0.027302 -0.024087       NaN  -0.004669   
Packet Length Max          -0.041388  0.041552       NaN  -0.069531   
Packet Length Mean         -0.009775  0.010685       NaN  -0.074391   
Packet Length Std          -0.052318  0.051171       NaN  -0.068782   
Packet Length Variance     -0.057375  0.055646       NaN  -0.065436   
FIN Flag Count              0.015090 -0.013024       NaN  -0.139347   
SYN Flag Count             -0.044921  0.295728       NaN   0.317159   
RST Flag Count             -0.398522  0.638621       NaN   0.210034   
PSH Flag Count              0.012537 -0.010753       NaN   0.003538   
ACK Flag Count              0.010968 -0.008731       NaN   0.007461   
URG Flag Count             -0.003177  0.003398       NaN   0.014874   
CWR Flag Count              0.001189  0.001479       NaN   0.007318   
ECE Flag Count             -0.002331  0.004930       NaN   0.011800   
Down/Up Ratio              -0.101601  0.203009       NaN   0.181389   
Average Packet Size        -0.006716  0.007738      